### ISSUE TITLE
Manual_CertificateSentMatchesCertificateReceived_Success failed on Ubuntu.14.04

### ISSUE THREAD

**Issue by mellinoe** [dotnet/corefx#9785]
Sample error: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/ubuntu14.04_debug_prtest/2465/testReport/junit/System.Net.Http.Functional.Tests/HttpClientHandler_ClientCertificates_Test/Manual_CertificateSentMatchesCertificateReceived_Success_numberOfRequests__3__reuseClient__True_/

Another: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/ubuntu14.04_debug_prtest/2303/testReport/junit/System.Net.Http.Functional.Tests/HttpClientHandler_ClientCertificates_Test/Manual_CertificateSentMatchesCertificateReceived_Success_numberOfRequests__3__reuseClient__True_/

```
MESSAGE:
System.Threading.Tasks.TaskCanceledException : A task was canceled.
+++++++++++++++++++
STACK TRACE:
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.CompilerServices.TaskAwaiter.GetResult() at System.Net.Http.Functional.Tests.HttpClientHandler_ClientCertificates_Test.<>c__DisplayClass5_0.<<Manual_CertificateSentMatchesCertificateReceived_Success>b__2>d.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.CompilerServices.TaskAwaiter.GetResult() at System.Net.Http.Functional.Tests.HttpClientHandler_ClientCertificates_Test.<>c__DisplayClass5_2.<<Manual_CertificateSentMatchesCertificateReceived_Success>b__4>d.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.CompilerServices.TaskAwaiter.GetResult() at System.Net.Test.Common.LoopbackServer.<>c__DisplayClass3_0.<CreateServerAsync>b__0(Task t) at System.Threading.Tasks.Task.Execute() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.CompilerServices.TaskAwaiter.GetResult() at System.Net.Http.Functional.Tests.HttpClientHandler_ClientCertificates_Test.<Manual_CertificateSentMatchesCertificateReceived_Success>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

```

**Comment by mellinoe** [comment]
I searched around for a dupe but didn't see anything directly related. Feel free to close as a dupe if there is a better issue in use.

**Comment by stephentoub** [comment]
It's unfortunately a general dupe of the same "A task has been canceled" that occurs sporadically across lots of the HTTP tests and across OSes.  I've stopped opening issues for them, but we need to solve it, as it's a reliability issue for CI in general.

**Comment by ericeil** [comment]
Unlike the other "a task was canceled" failures I've seen (e.g., #9270), this one seems to involve a test that executes entirely in-process, with no external service involved.  The only explicit server involved here is `LoopbackServer`.  I wonder, though, if there might be some external server involved indirectly?

**Comment by stephentoub** [comment]
>  I wonder, though, if there might be some external server involved indirectly?

There really shouldn't be.  I'm not sure what's going on here, but figuring this one out may shed some light on the others.

**Comment by ericeil** [comment]
I did some experiments yesterday to see if this might be due to ThreadPool starvation (just a random theory, probably biased by all the time I've spent working on the ThreadPool... :) ).  I found some minor starvation issues in the HTTP tests; I'll be submitting a PR to fix those.  But even with those fixes, I still occasionally saw the "a task was canceled" error in tests involving external servers, but never on the loopback server.  As I mentioned in #9270, today I don't see these failures *at all*, in any test.
